Fix issue #89; cannot use gen .pb.go from multiple services

hasAdamr opened the pull request:
Switch from outputting `third_party` imports to using
`google.golang.org/genproto` for `http.proto` and `annotations.proto`.

This means that all generated services will import the same `http.pb.go`
and `annotations.pb.go`, fixing the problem of the proto library not
allowing for extensions to be imported from different locations.

**UPDATE:**

Based on pull request comments, this pull has been updated to search all GOPATH's for `google.golang.org/genproto` in the `protoc` invocation. It has also been updated to use the `"go/build"` package for finding the directories of the ServicePackage and the PBPackage.

adamryman committed:
Fix issue #89; cannot use gen .pb.go from multiple services

Switch from outputting third_party imports to using
google.golang.org/genproto for the http.proto and annotations.proto

This means that all generated services will import the same http.pb.go
and annotations.pb.go, fixing the problem of the proto library not
allowing for extensions to be imported from different locations

hasIan commented on deftree/build_deftree.go:
```diff
@@ -92,30 +92,30 @@ func New(req *plugin.CodeGeneratorRequest, serviceFile io.Reader) (Deftree, erro
 }
 
-func NewFromString(def string) (Deftree, error) {
+func NewFromString(def, GOPATH string) (Deftree, error) {
```
I know `GOPATH` is special, but in this case it's just a parameter. It looks kind of weird to me to see a regular function parameter in all caps here.

hasAdamr replied:
In the go standard library, they treat the "$GOPATH" this way.

```
/usr/local/go/src/go/build/build.go
...
    GOOS        string // target operating system
    GOROOT      string // Go root
    GOPATH      string // Go path
    CgoEnabled bool // whether cgo can be used
```

https://github.com/golang/go/blob/master/src/go/build/build.go#L34

Though I just found this: https://github.com/golang/go/blob/master/src/cmd/go/build.go#L1790

```
for _, root := range gopath {
```

I could go with `gopath` but if it is attached to a struct I would like to use `GOPATH`

hasIan replied:
I don't feel too strongly either way; the mixed examples from the Go code itself satisfy me.

hasIan commented on deftree/build_deftree_test.go:
```diff
@@ -1,27 +1,34 @@
+var GOPATH string
+
+func init() {
+	GOPATH = filepath.SplitList(os.Getenv("GOPATH"))[0]
```
It would be polite to inform the user whether we're ignoring any of a multi-entry GOPATH. Like, hey bub, you've got four directories in your GOPATH and we're picking the first one and ignoring the rest. If someone is using a multi-entry GOPATH and this just silently truncates it, it could be a source of angst.

hasAdamr replied:
I just found this: https://github.com/golang/go/blob/master/src/cmd/go/build.go#L1790

```
for _, root := range gopath {
```

I could just range over the `gopath` list. This can be used throughout truss I believe. I can start rolling in this larger change in the morning.

hasIan replied:
Sounds good.

hasIan commented on README.md:
```diff
@@ -1,38 +1,60 @@
+2. Of course, install the Go compiler and tools from https://golang.org/.
+See https://golang.org/doc/install for details.
+3. Install the `protoc-gen-go` and `proto` packages for Go. The simplest 
+way is to run 
```
Looks like we need to add `go get google.golang.org/genproto/...` into this list of things to do before Truss can be used, yeah? Or, rather, before the output of Truss can be used.

hasAdamr replied:
That is right below this. 

``````
+   ```
+   go get -u google.golang.org/genproto    
+   ```
``````

hasIan replied:
@_@ I swear I read the list like three times and didn't see that. Losing it.

hasIan reviewed (commented):
Looks good! Minor comment about `GOPATH` looking kind of weird as a variable name, since it's not a constant, and Go doesn't even encourage all caps for constants. The `runtime` package is the only place I've really seen the all-caps thing before. Not a big deal, but it flaunts convention.

One comment about the setup documentation too, but that's it. Pretty much good to go, so far as I can tell.

adamryman committed:
Use all GOPATH's rather than just first

Using "go/build" for interacting with the GOPATH for the ServicePackage
and PBPackage.

Use all GOPATH's as imports for protoc to allow the user to install
google.golang.org/genproto in any GOPATH rather than just the first.

hasIan commented on truss/config.go:
```diff
@@ -1,35 +1,19 @@
-	GOPATH string
+	GoPath []string
 
 	// The go packge where .pb.go files protoc-gen-go creates will be written
```
"packge"

hasIan commented:
Looks good. Noticed a minor typo. :100: :shipit:
